zerver/lib/push_notifications.py

```diff
@@ -657,6 +664,18 @@ def get_remove_payload_gcm(
     gcm_options = {'priority': 'normal'}
     return gcm_payload, gcm_options
 
+def get_remove_payload_apns(user_profile: UserProfile, message_ids: List[int]) -> Dict[str, Any]:
+    zulip_data = get_base_payload(user_profile)
+    zulip_data.update({
+        'event': 'remove',
+        'zulip_message_ids': ','.join(str(id) for id in message_ids),
+    })
+    apns_data = {
+        'badge': get_apns_badge_counts(user_profile) - len(message_ids),
```

Comment:
I think this will double-count the messages in `message_ids`. Note the docstring on the caller, `handle_remove_push_notification`:

    """This should be called when a message that had previously had a
    mobile push executed is read.  [...]

So at this point these messages are already read, and should be out of the count in the first place.

Reply:
@gnprice thanks for the review. I didn't understand completely what do you mean here. Can you please elaborate on this?

Reply:
Looking closer, I think this code has approximately the right behavior. But this is one of the spots where I would have a lot more confidence in the first place if this PR added some appropriate unit tests! In particular:
> tests where [the badge count] gets decreased to a number above 0.

The thing I was concerned about is whether the messages in `message_ids` were included in the count made by `get_apns_badge_count`, or whether they'd be already excluded from that count by this point. I thought they might be already excluded, but from looking closer at the whole sequence of how we manage this `active_mobile_push_notification` flag, I see they're still included.

There's still a potential race condition here, in the case where for whatever reason we handle the same message ID through this codepath twice. Then the second time through, this subtraction will be inappropriate because the given message ID will not be part of the count on the left. So a better way to write this is to have the count be taken using `.exclude(message_id__in=[message_ids])`. See docs:
https://docs.djangoproject.com/en/3.0/topics/db/queries/#retrieving-specific-objects-with-filters
https://docs.djangoproject.com/en/3.0/ref/models/querysets/#in

For more details, #backend is the best place to ask.

Reply:
Thanks for the explanation. That was really helpful!